cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -0,0 +1,331 @@
+   * handles fetching, caching and unpacking the artifact.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch and unpack
+   * @return The Local Location of the directory that contains the unpacked artifact files
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching, caching or unpacking the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public File getAndUnpackArtifact(ArtifactId artifactId) throws Exception {
+    LOG.debug("Unpacking artifact {}", artifactId);
+    File jarLocation = getArtifact(artifactId);
+    File unpackDir = getUnpackLocalPath(artifactId, Long.valueOf(jarLocation.getName().split("\\.")[0]));
+    LOG.debug("Got unpack directory as {}", unpackDir);
+    if (!unpackDir.exists()) {
+      LOG.debug("Unpack directory doesn't exist, creating it now");
+      BundleJarUtil.unJar(jarLocation, unpackDir);
+    }
+    return unpackDir;
+  }
+
+  /**
+   * fetchArtifact attempts to connect to app fabric to download the given artifact. This method will throw {@link
+   * RetryableException} in certain circumstances so using this with the
+   */
+  private File fetchArtifact(ArtifactId artifactId,
+                             @Nullable Long lastModifiedTimestamp) throws IOException, ArtifactNotFoundException {
+
+    String namespaceId = artifactId.getNamespace();
+    ArtifactScope scope = ArtifactScope.USER;
+    // Cant use 'system' as the namespace in the request because that generates an error, the namespace doesnt matter
+    // as long as it exists. Using default because it will always be there
+    if (ArtifactScope.SYSTEM.toString().equalsIgnoreCase(namespaceId)) {
+      namespaceId = NamespaceId.DEFAULT.getEntityName();
+      scope = ArtifactScope.SYSTEM;
+    }
+
+    String url = String.format("namespaces/%s/artifacts/%s/versions/%s/download?scope=%s",
+                               namespaceId,
+                               artifactId.getArtifact(),
+                               artifactId.getVersion(),
+                               scope);
+
+    HttpURLConnection urlConn = remoteClient.openConnection(HttpMethod.GET, url);
+    if (lastModifiedTimestamp != null) {
+      LOG.debug("Found existing local version for {} with timestamp {}", artifactId, lastModifiedTimestamp);
+
+      ZonedDateTime lastModifiedDate = ZonedDateTime
+        .ofInstant(Instant.ofEpochMilli(lastModifiedTimestamp), ZoneId.of("GMT"));
+      urlConn.setRequestProperty(HttpHeaderNames.IF_MODIFIED_SINCE, lastModifiedDate.format(DATE_TIME_FORMATTER));
+    }
+
+    try {
+      // If we get this response that means we already have the most up to date artifact
+      if (urlConn.getResponseCode() == HttpURLConnection.HTTP_NOT_MODIFIED) {
+        LOG.debug("Call to app fabric returned NOT_MODIFIED");
```

Comment:
In log, always log with the contextual information. E.g. in this case, you should log with the artifactId, otherwise it will just be bunch of the same debug messages without carrying much value.

Reply:
done